deltachat-contact-tools/src/lib.rs

```diff
@@ -29,10 +29,116 @@ use std::fmt;
 use std::ops::Deref;
 
 use anyhow::bail;
+use anyhow::format_err;
+use anyhow::Context as _;
 use anyhow::Result;
+use chrono::DateTime;
 use once_cell::sync::Lazy;
 use regex::Regex;
 
+// TODOs to clean up:
+// - Check if sanitizing is done correctly everywhere
+// - Apply lints everywhere (https://doc.rust-lang.org/cargo/reference/workspaces.html#the-lints-table)
+
+#[derive(Debug)]
+/// A Contact, as represented in a VCard.
+pub struct VcardContact {
+    /// The email address, vcard property `email`
+    pub addr: String,
+    /// The contact's display name, vcard property `fn`
+    pub display_name: String,
+    /// The contact's public PGP key, vcard property `key`
+    pub key: Option<String>,
+    /// The contact's profile photo (=avatar), vcard property `photo`
+    pub profile_photo: Option<String>,
+    /// The timestamp when the vcard was created / last updated, vcard property `rev`
+    pub timestamp: Result<u64>,
+}
+
+pub fn parse_vcard(vcard: String) -> Result<Vec<VcardContact>> {
+    fn remove_prefix<'a>(s: &'a str, prefix: &str) -> Option<&'a str> {
+        let start_of_s = s.get(..prefix.len())?;
+
+        if start_of_s.eq_ignore_ascii_case(prefix) {
+            s.get(prefix.len()..)
+        } else {
+            None
+        }
+    }
+    fn vcard_property<'a>(s: &'a str, property: &str) -> Option<&'a str> {
+        let remainder = remove_prefix(s, property)?;
+
+        // TODO this doesn't handle the case where there are quotes around a colon
+        let (_params, value) = remainder.split_once(':')?;
+        Some(value)
+    }
+    fn parse_datetime(datetime: Option<&str>) -> Result<u64> {
+        let datetime = datetime.context("No timestamp in vcard")?;
```

Comment:
Maybe better to get rid of `Option`? Seems that an external check wouldn't be more complicated

Reply:
To achieve the exact same behavior without passing an `Option` here, we would probably need more LOC (at least I can't think of a nice way that would avoid it) because the `Result` is stored into the `timestamp` field, so we couldn't just use `?` at the caller site. But if you have a concrete idea how to improve the code (and still have it compile ofc), we can also go for that

Reply:
I checked, this works:
`let timestamp = datetime.context("No timestamp in vcard").and_then(parse_datetime);`